_posts/2025-04-25-from-recipes-to-vectors-building-a-hybrid-search-app-with-opensearch.md

```diff
@@ -0,0 +1,299 @@
+
+Specifically, when using vector fields for semantic search, you must specify the mappings and include a few special properties.
+For our dataset, the mapping is as follows:
+
+```
+PUT /recipes
+{
+  "settings": {
+    "index.knn": true
+  },
+  "mappings": {
+    "properties": {
+      "DatePublished": {
+        "type": "date"
+      },
+      "Name": {
+        "type": "text",
+        "fields": {
+          "keyword": { "type": "keyword" }}
+      },
+...
+      "RecipeCategory": {
+        "type": "keyword"
+      },
+      "RecipeInstructions": {
+        "type": "text"
+      },
+      "embedding": {
+        "type": "knn_vector",
+        "dimension": 1024
+      }
+    }
+  }
+}
+
+```
+
+Most of this should look familiar, but let's take a closer look at the few fields that are unique to OpenSearch and semantic search.
+
+### Understand index settings
+
+First of all, let's take a look at the settings field - it controls index specific configurations on how OpenSearch stores and manages the index. It is also the way to enable the OpenSearch [knn plugin](https://github.com/opensearch-project/k-NN) and enable semantic search.
+
+The most basic configuration would look like this:
+
+```
+  "settings": {
+    "index.knn": true,
+  }
+
+```
+
+* `index.knn: true` – The knn implementation is built as a plugin. It is enabled by default on most setups, but you still need to enable the knn plugin on the index by setting this setting.
+* `index.knn.algo_param.ef_construction` - Controls the number of bi-directional links per node in the HNSW graph. 
+Intuition tip: Higher values improve recall but increase memory usage.
+* `index.knn.algo_param.m` - Defines how many neighbors are considered during HNSW index construction. 
+Intuition tip: Higher values improve search accuracy but slow down indexing.
+* `index.knn.algo_param.ef_search` - Controls the accuracy of k-NN queries over HNSW indices. 
+Intuition tip: Higher values mean better accuracy, but slower search performance.
+* `index.knn.space_type` - The default vector space used to calculate the distance between vectors. The two most popular options are an `l2 space` (Euclidean distance) or `cosinesimil` (cosine similarity). 
+Intuition tip: Cosine similarity is usually great for semantic search and recommendations. Euclidean distance is often better for detecting similarity in images, face recognition, etc.
+
+Now let's explore the RecipeInstructions_embedding field, which is where we store vector embeddings for semantic search. By default the only two values you must specify are
+
+* `type: “knn_vector”` – To let OpenSearch know that this is a vector field.
+* `dimension: 1024` – This is the size of the vector and it would normally be defined by the embedding mode your our using. In our case, embed-english-v3 creates vectors with 1024 values.
```

Comment:
```suggestion
* `dimension: 1024` – This is the size of the vector, and it would normally be defined by the embedding mode you're using. In our case, `embed-english-v3` creates vectors with 1,024 values.
```